File: services/api/app/magic_pointer.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

OPEN_TAG = "[ACTIVE_CURSOR_CONTEXT]"
CLOSE_TAG = "[/ACTIVE_CURSOR_CONTEXT]"

_DEICTIC_INSTRUCTION = (
    "Deictic resolution: interpret 'this', 'that', 'it', 'here', and phrases like "
    "'this function' or 'this file' as referring to the ACTIVE_CURSOR_CONTEXT block below. "
    "Do not ask the user to paste code already shown in Selected Text or Cursor Line."
)

_ENTITY_RULES: list[tuple[str, re.Pattern[str], list[str]]] = [
    (
        "api_route",
        re.compile(r"@(?:app|router)\.(?:get|post|put|patch|delete|websocket)\([\"']([^\"']+)", re.I),
        [
            "Open the router module that registers this route",
            "Trace handler implementation and request/response models",
            "Run API tests covering this endpoint",
        ],
    ),
    (
        "http_path",
        re.compile(r"[\"'](/api/[^\"']+)[\"']"),
        [
            "Find FastAPI route or Next.js BFF handler for this path",
            "Check infra/gateway/nginx.conf proxy rules",
        ],
    ),
    (
        "npm_missing",
        re.compile(r"Cannot find module ['\"]([^'\"]+)['\"]"),
        ["Run npm install in the relevant workspace", "Add package to package.json"],
    ),
    (
        "pytest_failure",
        re.compile(r"(FAILED|AssertionError|pytest\.fail)", re.I),
        ["Run npm run api:test", "Apply Loop Engineering verify/fix protocol"],
    ),
    (
        "import_statement",
        re.compile(r"(?:from|import)\s+([\w.]+)"),
        ["Jump to module definition", "Verify dependency in requirements.txt or package.json"],
    ),
    (
        "terminal_error",
        re.compile(r"(Error:|Traceback \(most recent call last\)|npm ERR!|ELIFECYCLE)", re.I),
        ["Capture full stderr", "Run Loop Engineering on the failing command"],
    ),
]


@dataclass(frozen=True)
class DetectedEntity:
    kind: str
    value: str
    suggested_actions: list[str]


def detect_entities(text: str) -> list[DetectedEntity]:
    if not text.strip():
        return []

    found: list[DetectedEntity] = []
    seen: set[tuple[str, str]] = set()
    for kind, pattern, actions in _ENTITY_RULES:
        for match in pattern.finditer(text):
            value = next((group for group in match.groups() if group), match.group(0))
            key = (kind, value)
            if key in seen:
                continue
            seen.add(key)
            found.append(DetectedEntity(kind=kind, value=value, suggested_actions=actions))
    return found


def _clip(text: str, limit: int = 4000) -> str:
    if len(text) <= limit:
        return text
    return f"{text[: limit - 20]}\n...[truncated]"


def compose_active_cursor_context(
    *,
    file_path: str | None,
    line_number: int | None = None,
    selection_start_line: int | None = None,
    selection_end_line: int | None = None,
    selection_text: str | None = None,
    cursor_line_text: str | None = None,
    surrounding_context: str | None = None,
    include_deictic_instruction: bool = True,
) -> str:
    selection = (selection_text or "").strip()
    cursor_line = (cursor_line_text or "").strip()
    if not file_path and not selection and not cursor_line:
        return ""

    lines: list[str] = [OPEN_TAG]
    if include_deictic_instruction:
        lines.append(_DEICTIC_INSTRUCTION)

    if file_path:
        lines.append(f"File: {file_path}")
    if line_number:
        lines.append(f"Line: {line_number}")
    if selection_start_line and selection_end_line:
        lines.append(f"Selection Range: lines {selection_start_line}-{selection_end_line}")

    if selection:
        lines.append("Selected Text:")
        lines.append(_clip(selection))
    elif cursor_line:
        lines.append("Cursor Line:")
        lines.append(_clip(cursor_line))

    if surrounding_context and surrounding_context.strip():
        lines.append("Surrounding:")
        lines.append(_clip(surrounding_context.strip()))

    probe_text = selection or cursor_line or surrounding_context or ""
    entities = detect_entities(probe_text)
    if entities:
        lines.append("Detected Entities:")
        for entity in entities[:8]:
            actions = "; ".join(entity.suggested_actions[:2])
            lines.append(f"- {entity.kind}: {entity.value} → {actions}")

    lines.append(CLOSE_TAG)
    return "\n".join(lines)


def compose_from_message_context(context: dict[str, Any]) -> str:
    return compose_active_cursor_context(
        file_path=context.get("current_file"),
        line_number=context.get("line_number"),
        selection_start_line=context.get("selection_start_line"),
        selection_end_line=context.get("selection_end_line"),
        selection_text=context.get("selection_text"),
        cursor_line_text=context.get("cursor_line_text"),
        surrounding_context=context.get("surrounding_context"),
    )


def serialize_entities(context: dict[str, Any]) -> list[dict[str, Any]]:
    probe = " ".join(
        filter(
            None,
            [
                context.get("selection_text"),
                context.get("cursor_line_text"),
                context.get("surrounding_context"),
            ],
        )
    )
    return [
        {"kind": item.kind, "value": item.value, "suggested_actions": item.suggested_actions}
        for item in detect_entities(probe)
    ]

File: packages/shared/src/magic-pointer.js
const OPEN_TAG = "[ACTIVE_CURSOR_CONTEXT]";
const CLOSE_TAG = "[/ACTIVE_CURSOR_CONTEXT]";
const STORAGE_KEY = "codeforge_magic_pointer";

const DEICTIC_RULE =
  "Deictic resolution: interpret 'this', 'that', 'it', 'here', and phrases like " +
  "'this function' or 'this file' as referring to the ACTIVE_CURSOR_CONTEXT block below.";

/**
 * @param {object} params
 * @param {string} [params.filePath]
 * @param {number} [params.lineNumber]
 * @param {number} [params.selectionStartLine]
 * @param {number} [params.selectionEndLine]
 * @param {string} [params.selectionText]
 * @param {string} [params.cursorLineText]
 * @param {string} [params.surroundingContext]
 */
export function formatActiveCursorContext({
  filePath,
  lineNumber,
  selectionStartLine,
  selectionEndLine,
  selectionText,
  cursorLineText,
  surroundingContext,
}) {
  const selection = (selectionText || "").trim();
  const cursorLine = (cursorLineText || "").trim();
  if (!filePath && !selection && !cursorLine) {
    return "";
  }

  const lines = [OPEN_TAG, DEICTIC_RULE];
  if (filePath) {
    lines.push(`File: ${filePath}`);
  }
  if (lineNumber) {
    lines.push(`Line: ${lineNumber}`);
  }
  if (selectionStartLine && selectionEndLine) {
    lines.push(`Selection Range: lines ${selectionStartLine}-${selectionEndLine}`);
  }
  if (selection) {
    lines.push("Selected Text:", selection.slice(0, 4000));
  } else if (cursorLine) {
    lines.push("Cursor Line:", cursorLine.slice(0, 4000));
  }
  if (surroundingContext?.trim()) {
    lines.push("Surrounding:", surroundingContext.trim().slice(0, 4000));
  }
  lines.push(CLOSE_TAG);
  return lines.join("\n");
}

/**
 * Build surrounding lines with line numbers; `>` marks selection/cursor band.
 */
export function buildSurroundingContext(content, startLine, endLine, radius = 5) {
  if (!content) {
    return "";
  }
  const lines = content.split("\n");
  const start = Math.max(0, startLine - 1 - radius);
  const end = Math.min(lines.length, endLine + radius);
  return lines
    .slice(start, end)
    .map((line, index) => {
      const lineNo = start + index + 1;
      const marker = lineNo >= startLine && lineNo <= endLine ? ">" : " ";
      return `${marker} ${String(lineNo).padStart(4)}| ${line}`;
    })
    .join("\n");
}

export function persistMagicPointerState(payload) {
  if (typeof sessionStorage === "undefined") {
    return;
  }
  try {
    sessionStorage.setItem(STORAGE_KEY, JSON.stringify({ ...payload, updated_at: Date.now() }));
  } catch {
    // ignore quota errors
  }
}

export function readMagicPointerState() {
  if (typeof sessionStorage === "undefined") {
    return null;
  }
  try {
    const raw = sessionStorage.getItem(STORAGE_KEY);
    return raw ? JSON.parse(raw) : null;
  } catch {
    return null;
  }
}

const ENTITY_RULES = [
  {
    kind: "api_route",
    pattern: /@(?:app|router)\.(?:get|post|put|patch|delete|websocket)\(["']([^"']+)/i,
    actions: ["Trace router handler", "Run API tests"],
  },
  {
    kind: "http_path",
    pattern: /["'](\/api\/[^"']+)["']/,
    actions: ["Find BFF or FastAPI route", "Check gateway nginx"],
  },
  {
    kind: "npm_missing",
    pattern: /Cannot find module ['"]([^'"]+)['"]/,
    actions: ["npm install in workspace", "Add to package.json"],
  },
  {
    kind: "pytest_failure",
    pattern: /(FAILED|AssertionError|pytest\.fail)/i,
    actions: ["npm run api:test", "Loop Engineering fix"],
  },
  {
    kind: "import_statement",
    pattern: /(?:from|import)\s+([\w.]+)/,
    actions: ["Go to definition", "Check manifest"],
  },
  {
    kind: "terminal_error",
    pattern: /(Error:|Traceback \(most recent call last\)|npm ERR!)/i,
    actions: ["Capture stderr", "Run verify loop"],
  },
];

/**
 * @param {string} text
 * @returns {{ kind: string, value: string, suggested_actions: string[] }[]}
 */
export function detectEntitiesInText(text) {
  if (!text?.trim()) {
    return [];
  }
  const found = [];
  const seen = new Set();
  for (const rule of ENTITY_RULES) {
    const matches = text.matchAll(new RegExp(rule.pattern.source, rule.pattern.flags + "g"));
    for (const match of matches) {
      const value = match[1] || match[0];
      const key = `${rule.kind}:${value}`;
      if (seen.has(key)) {
        continue;
      }
      seen.add(key);
      found.push({ kind: rule.kind, value, suggested_actions: rule.actions });
    }
  }
  return found.slice(0, 8);
}

/**
 * @param {object} params
 * @param {string} [params.filePath]
 * @param {string} [params.content]
 * @param {{ startLine?: number, endLine?: number, text?: string } | null} [params.selection]
 * @param {{ lineNumber?: number } | null} [params.cursor]
 * @param {string} [params.hoverToken]
 */
export function buildPointerPayload({ filePath, content = "", selection, cursor, hoverToken }) {
  const lineNumber = cursor?.lineNumber || selection?.startLine || 1;
  const lines = content.split("\n");
  const cursorLineText = lines[Math.max(0, lineNumber - 1)] || "";
  const bandStart = selection?.startLine || lineNumber;
  const bandEnd = selection?.endLine || lineNumber;
  const probe = [selection?.text, cursorLineText, hoverToken].filter(Boolean).join("\n");
  return {
    file_path: filePath || null,
    line_number: lineNumber,
    selection_start_line: selection?.startLine ?? null,
    selection_end_line: selection?.endLine ?? null,
    selection_text: selection?.text ?? null,
    cursor_line_text: cursorLineText,
    hover_token: hoverToken || null,
    surrounding_context: buildSurroundingContext(content, bandStart, bandEnd, 5),
    detected_entities: detectEntitiesInText(probe),
  };
}

export { OPEN_TAG, CLOSE_TAG, STORAGE_KEY };
